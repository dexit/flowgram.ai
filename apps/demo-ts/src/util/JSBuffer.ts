export class JSBuffer {
  buffer: string[] = []

  constructor(public indent = '') {}

  clear(): JSBuffer {
    this.buffer = []
    return this
  }

  push(...args: string[]): JSBuffer {
    this.buffer.push(...args)
    return this
  }

  pushLine(...args: string[]): JSBuffer {
    this.buffer.push(...args, '\n')
    return this
  }

  toJs(separator: string = ''): string {
    // TODO js string perf
    // FIXME last line indent
    return (
      this.buffer
        .join(separator)
        .split('\n')
        .map((line) => `${this.indent}${line}`)
        // .map((line) => (line !== '\n' ? `${this.indent}${line}` : line))
        .join('\n')
    )
  }

  toString(): string {
    return this.toJs()
  }
}
