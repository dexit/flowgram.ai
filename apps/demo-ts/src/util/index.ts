export * from './log'
export * from './JSBuffer'
