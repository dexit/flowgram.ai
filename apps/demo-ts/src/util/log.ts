import debug from 'debug'

export const info = debug('log:info')
export const log = info
export const warn = debug('log:warn')
export const error = debug('log:error')
