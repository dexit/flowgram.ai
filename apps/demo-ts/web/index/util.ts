
export const codeMirrorBaseConfig: CodeMirror.EditorConfiguration = Object.freeze({
  mode: { name: 'javascript', json: true },
  lineNumbers: true,
  lineWrapping: false,
  theme: 'twilight',
  foldGutter: true,
  gutters: ['CodeMirror-linenumbers', 'CodeMirror-foldgutter'],
  readOnly: true,
  cursorBlinkRate: -1,
})