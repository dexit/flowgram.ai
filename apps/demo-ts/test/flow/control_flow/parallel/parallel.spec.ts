import { runCases } from '../../util'

runCases(
  ['parallel_basic', 'parallel_nest', 'parallel_outputs', 'parallel_outputs_full'],
  __dirname
)
