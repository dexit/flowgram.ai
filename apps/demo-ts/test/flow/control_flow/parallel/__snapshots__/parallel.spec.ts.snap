// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flow parallel_basic | {} 1`] = `
"export async function __main(__args) {
  //=== start@start | {level:1,blockIndex:0}
  const start = {}
  start.outputs = {}
  Object.assign(start.outputs, __args)
  //=== variable@variable | {level:1,blockIndex:1}
  const variable = {}
  variable.outputs = {
    a: '',
  }
  variable.outputs.a = '0'
  //=== parallel@parallel | {level:1,blockIndex:2}
  const parallel = {}
  parallel.outputs = {}
  await Promise.all([
    new Promise(async (__resolve, __reject) => {
      //=== branch_1@block | {level:2,blockIndex:0,parent:'parallel'}
      const branch_1 = {}
      branch_1.outputs = {}
      //=== assign_1@variable | {level:3,blockIndex:0,parent:'branch_1'}
      const assign_1 = {}
      assign_1.outputs = {}
      variable.outputs.a = '1'
      branch_1.outputs = assign_1.outputs
      parallel.outputs.branch_1 = branch_1
      __resolve()
    }), // branch_1@block
    new Promise(async (__resolve, __reject) => {
      //=== branch_2@block | {level:2,blockIndex:1,parent:'parallel'}
      const branch_2 = {}
      branch_2.outputs = {}
      //=== assign_2@variable | {level:3,blockIndex:0,parent:'branch_2'}
      const assign_2 = {}
      assign_2.outputs = {}
      variable.outputs.a = '2'
      branch_2.outputs = assign_2.outputs
      parallel.outputs.branch_2 = branch_2
      __resolve()
    }), // branch_2@block
    new Promise(async (__resolve, __reject) => {
      //=== branch_3@block | {level:2,blockIndex:2,parent:'parallel'}
      const branch_3 = {}
      branch_3.outputs = {}
      //=== assign_3@variable | {level:3,blockIndex:0,parent:'branch_3'}
      const assign_3 = {}
      assign_3.outputs = {}
      variable.outputs.a = '3'
      branch_3.outputs = assign_3.outputs
      parallel.outputs.branch_3 = branch_3
      __resolve()
    }), // branch_3@block
  ]) // parallel@parallel
  //=== end@end | {level:1,blockIndex:3}
  const end = {}
  end.data = {
    result: variable.outputs.a,
  }
  return end.data
}
"
`;

exports[`flow parallel_basic | {} 2`] = `
{
  "error": {
    "code": "",
    "msg": "",
  },
  "result": "3",
}
`;

exports[`flow parallel_nest | {} 1`] = `
"export async function __main(__args) {
  //=== start@start | {level:1,blockIndex:0}
  const start = {}
  start.outputs = {}
  Object.assign(start.outputs, __args)
  //=== parallel_0@parallel | {level:1,blockIndex:1}
  const parallel_0 = {}
  parallel_0.outputs = {}
  await Promise.all([
    new Promise(async (__resolve, __reject) => {
      //=== parallel_0_branch_0@block | {level:2,blockIndex:0,parent:'parallel_0'}
      const parallel_0_branch_0 = {}
      parallel_0_branch_0.outputs = {}
      //=== script_0@script | {level:3,blockIndex:0,parent:'parallel_0_branch_0'}
      const script_0 = {}
      script_0.outputs = {
        output: '',
      }
      script_0.data = {}
      const __func_script_0 = (inputs, outputs) => {
        outputs.output = inputs.name
      }
      __func_script_0(script_0.data, script_0.outputs)
      parallel_0_branch_0.outputs = script_0.outputs
      parallel_0.outputs.parallel_0_branch_0 = parallel_0_branch_0
      __resolve()
    }), // parallel_0_branch_0@block
    new Promise(async (__resolve, __reject) => {
      //=== parallel_0_branch_1@block | {level:2,blockIndex:1,parent:'parallel_0'}
      const parallel_0_branch_1 = {}
      parallel_0_branch_1.outputs = {}
      //=== parallel_1@parallel | {level:3,blockIndex:0,parent:'parallel_0_branch_1'}
      const parallel_1 = {}
      parallel_1.outputs = {}
      await Promise.all([
        new Promise(async (__resolve, __reject) => {
          //=== parallel_1_branch_0@block | {level:4,blockIndex:0,parent:'parallel_1'}
          const parallel_1_branch_0 = {}
          parallel_1_branch_0.outputs = {}
          //=== script_1@script | {level:5,blockIndex:0,parent:'parallel_1_branch_0'}
          const script_1 = {}
          script_1.outputs = {
            output: '',
          }
          script_1.data = {}
          const __func_script_1 = (inputs, outputs) => {
            outputs.output = inputs.name
          }
          __func_script_1(script_1.data, script_1.outputs)
          parallel_1_branch_0.outputs = script_1.outputs
          parallel_1.outputs.parallel_1_branch_0 = parallel_1_branch_0
          __resolve()
        }), // parallel_1_branch_0@block
        new Promise(async (__resolve, __reject) => {
          //=== parallel_1_branch_1@block | {level:4,blockIndex:1,parent:'parallel_1'}
          const parallel_1_branch_1 = {}
          parallel_1_branch_1.outputs = {}
          //=== parallel_2@parallel | {level:5,blockIndex:0,parent:'parallel_1_branch_1'}
          const parallel_2 = {}
          parallel_2.outputs = {}
          await Promise.all([
            new Promise(async (__resolve, __reject) => {
              //=== parallel_2_branch_0@block | {level:6,blockIndex:0,parent:'parallel_2'}
              const parallel_2_branch_0 = {}
              parallel_2_branch_0.outputs = {}
              //=== script_3@script | {level:7,blockIndex:0,parent:'parallel_2_branch_0'}
              const script_3 = {}
              script_3.outputs = {
                output: '',
              }
              script_3.data = {}
              const __func_script_3 = (inputs, outputs) => {
                outputs.output = inputs.name
              }
              __func_script_3(script_3.data, script_3.outputs)
              parallel_2_branch_0.outputs = script_3.outputs
              parallel_2.outputs.parallel_2_branch_0 = parallel_2_branch_0
              __resolve()
            }), // parallel_2_branch_0@block
            new Promise(async (__resolve, __reject) => {
              //=== parallel_2_branch_1@block | {level:6,blockIndex:1,parent:'parallel_2'}
              const parallel_2_branch_1 = {}
              parallel_2_branch_1.outputs = {}
              //=== script_4@script | {level:7,blockIndex:0,parent:'parallel_2_branch_1'}
              const script_4 = {}
              script_4.outputs = {
                output: '',
              }
              script_4.data = {}
              const __func_script_4 = (inputs, outputs) => {
                outputs.output = inputs.name
              }
              __func_script_4(script_4.data, script_4.outputs)
              parallel_2_branch_1.outputs = script_4.outputs
              parallel_2.outputs.parallel_2_branch_1 = parallel_2_branch_1
              __resolve()
            }), // parallel_2_branch_1@block
          ]) // parallel_2@parallel
          //=== observer_1@script | {level:5,blockIndex:1,parent:'parallel_1_branch_1'}
          const observer_1 = {}
          observer_1.outputs = {
            output: '',
          }
          observer_1.data = {}
          const __func_observer_1 = (inputs, outputs) => {
            outputs.output = inputs.name
          }
          __func_observer_1(observer_1.data, observer_1.outputs)
          parallel_1_branch_1.outputs = observer_1.outputs
          parallel_1.outputs.parallel_1_branch_1 = parallel_1_branch_1
          __resolve()
        }), // parallel_1_branch_1@block
      ]) // parallel_1@parallel
      //=== observer_2@script | {level:3,blockIndex:1,parent:'parallel_0_branch_1'}
      const observer_2 = {}
      observer_2.outputs = {
        output: '',
      }
      observer_2.data = {}
      const __func_observer_2 = (inputs, outputs) => {
        outputs.output = inputs.name
      }
      __func_observer_2(observer_2.data, observer_2.outputs)
      parallel_0_branch_1.outputs = observer_2.outputs
      parallel_0.outputs.parallel_0_branch_1 = parallel_0_branch_1
      __resolve()
    }), // parallel_0_branch_1@block
  ]) // parallel_0@parallel
  //=== observer_3@script | {level:1,blockIndex:2}
  const observer_3 = {}
  observer_3.outputs = {
    output: '',
  }
  observer_3.data = {}
  const __func_observer_3 = (inputs, outputs) => {
    outputs.output = inputs.name
  }
  __func_observer_3(observer_3.data, observer_3.outputs)
  //=== end@end | {level:1,blockIndex:3}
  const end = {}
  end.data = {}
  return end.data
}
"
`;

exports[`flow parallel_nest | {} 2`] = `
{
  "error": {
    "code": "",
    "msg": "",
  },
}
`;

exports[`flow parallel_outputs | {} 1`] = `
"export async function __main(__args) {
  //=== start@start | {level:1,blockIndex:0}
  const start = {}
  start.outputs = {}
  Object.assign(start.outputs, __args)
  //=== parallel@parallel | {level:1,blockIndex:1}
  const parallel = {}
  parallel.outputs = {
    branch_0: {
      outputs: {
        a: '',
      },
    },
    branch_1: {
      outputs: {
        parallel_2_branch_0: {
          outputs: {
            a: '',
          },
        },
        parallel_2_branch_1: {
          outputs: {
            a: '',
          },
        },
      },
    },
  }
  await Promise.all([
    new Promise(async (__resolve, __reject) => {
      //=== branch_0@block | {level:2,blockIndex:0,parent:'parallel'}
      const branch_0 = {}
      branch_0.outputs = {
        a: '',
      }
      //=== v_0@variable | {level:3,blockIndex:0,parent:'branch_0'}
      const v_0 = {}
      v_0.outputs = {
        a: '',
      }
      v_0.outputs.a = '1'
      branch_0.outputs = v_0.outputs
      parallel.outputs.branch_0 = branch_0
      __resolve()
    }), // branch_0@block
    new Promise(async (__resolve, __reject) => {
      //=== branch_1@block | {level:2,blockIndex:1,parent:'parallel'}
      const branch_1 = {}
      branch_1.outputs = {
        parallel_2_branch_0: {
          outputs: {
            a: '',
          },
        },
        parallel_2_branch_1: {
          outputs: {
            a: '',
          },
        },
      }
      //=== v_1@variable | {level:3,blockIndex:0,parent:'branch_1'}
      const v_1 = {}
      v_1.outputs = {
        a: '',
      }
      v_1.outputs.a = '2'
      //=== parallel_2@parallel | {level:3,blockIndex:1,parent:'branch_1'}
      const parallel_2 = {}
      parallel_2.outputs = {
        parallel_2_branch_0: {
          outputs: {
            a: '',
          },
        },
        parallel_2_branch_1: {
          outputs: {
            a: '',
          },
        },
      }
      await Promise.all([
        new Promise(async (__resolve, __reject) => {
          //=== parallel_2_branch_0@block | {level:4,blockIndex:0,parent:'parallel_2'}
          const parallel_2_branch_0 = {}
          parallel_2_branch_0.outputs = {
            a: '',
          }
          //=== v2@variable | {level:5,blockIndex:0,parent:'parallel_2_branch_0'}
          const v2 = {}
          v2.outputs = {
            a: '',
          }
          v2.outputs.a = '3'
          parallel_2_branch_0.outputs = v2.outputs
          parallel_2.outputs.parallel_2_branch_0 = parallel_2_branch_0
          __resolve()
        }), // parallel_2_branch_0@block
        new Promise(async (__resolve, __reject) => {
          //=== parallel_2_branch_1@block | {level:4,blockIndex:1,parent:'parallel_2'}
          const parallel_2_branch_1 = {}
          parallel_2_branch_1.outputs = {
            a: '',
          }
          //=== v3@variable | {level:5,blockIndex:0,parent:'parallel_2_branch_1'}
          const v3 = {}
          v3.outputs = {
            a: '',
          }
          v3.outputs.a = '4'
          parallel_2_branch_1.outputs = v3.outputs
          parallel_2.outputs.parallel_2_branch_1 = parallel_2_branch_1
          __resolve()
        }), // parallel_2_branch_1@block
      ]) // parallel_2@parallel
      branch_1.outputs = parallel_2.outputs
      parallel.outputs.branch_1 = branch_1
      __resolve()
    }), // branch_1@block
  ]) // parallel@parallel
  //=== end@end | {level:1,blockIndex:2}
  const end = {}
  end.data = {
    res: parallel.outputs,
  }
  return end.data
}
"
`;

exports[`flow parallel_outputs | {} 2`] = `
{
  "error": {
    "code": "",
    "msg": "",
  },
  "res": {
    "branch_0": {
      "outputs": {
        "a": "1",
      },
    },
    "branch_1": {
      "outputs": {
        "parallel_2_branch_0": {
          "outputs": {
            "a": "3",
          },
        },
        "parallel_2_branch_1": {
          "outputs": {
            "a": "4",
          },
        },
      },
    },
  },
}
`;

exports[`flow parallel_outputs_full | {} 1`] = `
"export async function __main(__args) {
  //=== start@start | {level:1,blockIndex:0}
  const start = {}
  start.outputs = {}
  Object.assign(start.outputs, __args)
  //=== parallel@parallel | {level:1,blockIndex:1}
  const parallel = {}
  parallel.outputs = {
    branch_0: {
      outputs: {
        a: '',
      },
    },
    branch_1: {
      outputs: {
        parallel_2_branch_0: {
          outputs: {
            a: '',
          },
        },
        parallel_2_branch_1: {
          outputs: {
            a: '',
          },
        },
      },
    },
    branch_2: {
      outputs: {},
    },
    branch_3: {
      outputs: {},
    },
    branch_4: {
      outputs: {},
    },
  }
  await Promise.all([
    new Promise(async (__resolve, __reject) => {
      //=== branch_0@block | {level:2,blockIndex:0,parent:'parallel'}
      const branch_0 = {}
      branch_0.outputs = {
        a: '',
      }
      //=== v_0@variable | {level:3,blockIndex:0,parent:'branch_0'}
      const v_0 = {}
      v_0.outputs = {
        a: '',
      }
      v_0.outputs.a = '1'
      branch_0.outputs = v_0.outputs
      parallel.outputs.branch_0 = branch_0
      __resolve()
    }), // branch_0@block
    new Promise(async (__resolve, __reject) => {
      //=== branch_1@block | {level:2,blockIndex:1,parent:'parallel'}
      const branch_1 = {}
      branch_1.outputs = {
        parallel_2_branch_0: {
          outputs: {
            a: '',
          },
        },
        parallel_2_branch_1: {
          outputs: {
            a: '',
          },
        },
      }
      //=== v_1@variable | {level:3,blockIndex:0,parent:'branch_1'}
      const v_1 = {}
      v_1.outputs = {
        a: '',
      }
      v_1.outputs.a = '2'
      //=== parallel_2@parallel | {level:3,blockIndex:1,parent:'branch_1'}
      const parallel_2 = {}
      parallel_2.outputs = {
        parallel_2_branch_0: {
          outputs: {
            a: '',
          },
        },
        parallel_2_branch_1: {
          outputs: {
            a: '',
          },
        },
      }
      await Promise.all([
        new Promise(async (__resolve, __reject) => {
          //=== parallel_2_branch_0@block | {level:4,blockIndex:0,parent:'parallel_2'}
          const parallel_2_branch_0 = {}
          parallel_2_branch_0.outputs = {
            a: '',
          }
          //=== v2@variable | {level:5,blockIndex:0,parent:'parallel_2_branch_0'}
          const v2 = {}
          v2.outputs = {
            a: '',
          }
          v2.outputs.a = '3'
          parallel_2_branch_0.outputs = v2.outputs
          parallel_2.outputs.parallel_2_branch_0 = parallel_2_branch_0
          __resolve()
        }), // parallel_2_branch_0@block
        new Promise(async (__resolve, __reject) => {
          //=== parallel_2_branch_1@block | {level:4,blockIndex:1,parent:'parallel_2'}
          const parallel_2_branch_1 = {}
          parallel_2_branch_1.outputs = {
            a: '',
          }
          //=== v3@variable | {level:5,blockIndex:0,parent:'parallel_2_branch_1'}
          const v3 = {}
          v3.outputs = {
            a: '',
          }
          v3.outputs.a = '4'
          parallel_2_branch_1.outputs = v3.outputs
          parallel_2.outputs.parallel_2_branch_1 = parallel_2_branch_1
          __resolve()
        }), // parallel_2_branch_1@block
      ]) // parallel_2@parallel
      branch_1.outputs = parallel_2.outputs
      parallel.outputs.branch_1 = branch_1
      __resolve()
    }), // branch_1@block
    new Promise(async (__resolve, __reject) => {
      //=== branch_2@block | {level:2,blockIndex:2,parent:'parallel'}
      const branch_2 = {}
      branch_2.outputs = {}
      //=== batch@batch | {level:3,blockIndex:0,parent:'branch_2'}
      const batch = {}
      batch.outputs = {}
      batch.localVariables = {
        item: 0,
        index: 0,
      }
      // concurrency: 1
      batch.__array = [1, 2, 3]
      batch.__array.forEach((__item, __index) => {
        batch.localVariables.item = __item
        batch.localVariables.index = __index
        //=== variable_empty@variable | {level:4,blockIndex:0,parent:'batch'}
        const variable_empty = {}
        variable_empty.outputs = {}
      }) // batch@batch
      branch_2.outputs = batch.outputs
      parallel.outputs.branch_2 = branch_2
      __resolve()
    }), // branch_2@block
    new Promise(async (__resolve, __reject) => {
      //=== branch_3@block | {level:2,blockIndex:3,parent:'parallel'}
      const branch_3 = {}
      branch_3.outputs = {}
      //=== loop@forloop | {level:3,blockIndex:0,parent:'branch_3'}
      const loop = {}
      loop.outputs = {}
      loop.localVariables = {
        i: 0,
      }
      for (
        loop.localVariables.i = 0;
        loop.localVariables.i < 3;
        loop.localVariables.i = loop.localVariables.i + 1
      ) {
        //=== variable_empty_1@variable | {level:4,blockIndex:0,parent:'loop'}
        const variable_empty_1 = {}
        variable_empty_1.outputs = {}
      } // loop@forloop
      branch_3.outputs = loop.outputs
      parallel.outputs.branch_3 = branch_3
      __resolve()
    }), // branch_3@block
    new Promise(async (__resolve, __reject) => {
      //=== branch_4@block | {level:2,blockIndex:4,parent:'parallel'}
      const branch_4 = {}
      branch_4.outputs = {}
      //=== condition@condition | {level:3,blockIndex:0,parent:'branch_4'}
      const condition = {}
      condition.outputs = {}
      //=== condition_case_1@case | {level:3,blockIndex:0,parent:'condition'}
      if (true) {
        //=== variable_empty_2@variable | {level:4,blockIndex:0,parent:'condition_case_1'}
        const variable_empty_2 = {}
        variable_empty_2.outputs = {}
      } // condition_case_1@case
      //=== condition_default@default | {level:3,blockIndex:1,parent:'condition'}
      else {
        //=== variable_empty_3@variable | {level:4,blockIndex:0,parent:'condition_default'}
        const variable_empty_3 = {}
        variable_empty_3.outputs = {}
      } // condition_default@default
      branch_4.outputs = condition.outputs
      parallel.outputs.branch_4 = branch_4
      __resolve()
    }), // branch_4@block
    new Promise(async (__resolve, __reject) => {
      //=== branch_5@block | {level:2,blockIndex:5,parent:'parallel'}
      const branch_5 = {}
      branch_5.outputs = {}
      parallel.outputs.branch_5 = branch_5
      __resolve()
    }), // branch_5@block
  ]) // parallel@parallel
  //=== end@end | {level:1,blockIndex:2}
  const end = {}
  end.data = {
    res: parallel.outputs,
  }
  return end.data
}
"
`;

exports[`flow parallel_outputs_full | {} 2`] = `
{
  "error": {
    "code": "",
    "msg": "",
  },
  "res": {
    "branch_0": {
      "outputs": {
        "a": "1",
      },
    },
    "branch_1": {
      "outputs": {
        "parallel_2_branch_0": {
          "outputs": {
            "a": "3",
          },
        },
        "parallel_2_branch_1": {
          "outputs": {
            "a": "4",
          },
        },
      },
    },
    "branch_2": {
      "outputs": {},
    },
    "branch_3": {
      "outputs": {},
    },
    "branch_4": {
      "outputs": {},
    },
    "branch_5": {
      "outputs": {},
    },
  },
}
`;
